Expose GEOS Hausdorff distance calculations to QgsGeometry, add expression function

nyalldawson opened the pull request.

nyalldawson committed:
Expose GEOS Hausdorff distance calculations to QgsGeometry

nyalldawson committed:
[FEATURE] Expression function for hausdorff distance

Allows calculation of the Hausdorff distance between two
geometries

nyalldawson committed:
Fix evaluation of expression functions which use optional arguments
with null default values

m-kuhn commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -1447,6 +1447,28 @@ double QgsGeometry::distance( const QgsGeometry &geom ) const
+  }
+
+  QgsGeos g( d->geometry );
+  return g.hausdorffDistance( geom.d->geometry );
```
What about a bit of error reporting?

```cpp
return g.hausdorffDistance( geom.d->geometry, d->error );
```

nyalldawson replied:
Wouldn't that require a detach?

m-kuhn replied:
Right.
What do you think about switching the `QString* errorMsg = nullptr` param to a `std::function < void( const QString& error ) > errorCallback = nullptr`?

nyalldawson replied:
Hmm. Sounds a bit complex to use. Maybe we should just SIP_OUT a string pointer and leave it at that?

m-kuhn replied:
I think the SIP_OUT approach sounds complex to use, it will require creating a new QString and passing it into QgsGeometry functions every single time you call them in C++. I'd prefer to have all this handling hidden away in QgsGeometry, that's why I introduced `QgsGeometry::error()` which should either be used always or never.

nyalldawson replied:
Alternative idea: we move the error string out of the private geometry class and it becomes a mutable member just of QgsGeometry, not copied when QgsGeometry is copied.

m-kuhn replied:
Is there any risk involved with this combined with multithreading?

m-kuhn replied:
Forget that. Better question.

Will this work with

```cpp
QgsGeometry QgsGeometry::buffer( double distance )
{
   QgsGeometry result;
   result.mError = tr( "Something has gone wrong" );
   return result;
}
```

nyalldawson replied:
Ok, it'll need to be copied too.

nyalldawson committed:
Geometry error string is not shared

Instead move it to a private QgsGeometry member. It's copied
with QgsGeometry, but not the underlying private geometry
data.

This means the error string can be changed without detaching
(and copying) the geometry data.

Also make greater use of error string for all use of GEOS
functions.

nyalldawson commented:
@m-kuhn how's that for you?

nyalldawson commented:
Some answers in advance to questions you may have:

1. I've made it so methods which return a new geometry will have any error set on both the original AND the returned geometry. This means that both

```
p = geometry.centroid()
if not p:
   print(p.lastError())
```
and

    p = geometry.centroid()
    if not p:
       print(geometry.lastError())

are both valid. To me this is the correct approach.

2. I always clear the last error string when a new operation is performed. While this means that:

```
c = geometry.centroid() # some error occurs here
b = c.buffer(100)
if not b:
   print(c.lastError())
```

will not work (since the error from the centroid is cleared by the buffer call), I think that's acceptable. If we go the other way and leave the error around, then we'd get this undesirable behavior:

```
distance = geometry.distance(other_geometry) # some error happens here
... # do some other unrelated things
c = geometry.centroid()
...
print(c.lastError()) # lastError here is the error from the unrelated distance call on geometry
```

I prefer the clearing approach - it encourages writing safer code like
```
c = geometry.centroid() # some error occurs here
if not c:
   print(c.lastError())

b = c.buffer(100)
if not b:
   print(c.lastError())
```

m-kuhn commented:
Looks like a clean solution.

I was also wondering about error reporting of methods returning a geometry (where I expect the errors in the returned geom) compared to methods returning a value (where we need to resort to the parent geom), so reporting in both makes sense.

The only even cleaner solution would be exceptions which we can do in C++ for QGIS 4, when we have safe RAII code all over the place :)
I wonder if that could be done for the python part already now, but I'm a bit afraid that this will either be custom python code for each single method or a considerably complex generic solution.

nyalldawson commented:
> The only even cleaner solution would be exceptions which we can do in C++ for QGIS 4, when we have safe RAII code all over the place :)
I wonder if that could be done for the python part already now, but I'm a bit afraid that this will either be custom python code for each single method or a considerably complex generic solution.

Exception handling between c++/python is also really tricky. I'm hitting an issue with processing where an uncaught c++ exception thrown during Python code can hang QGIS when it occurs in a background thread (somewhere the Python global lock isn't being released, but i'm struggling to work out where).

I think this is a good compromise for now :)

m-kuhn commented:
> Exception handling between c++/python is also really tricky. I'm hitting an issue with processing where an uncaught c++ exception thrown during Python code can hang QGIS when it occurs in a background thread (somewhere the Python global lock isn't being released, but i'm struggling to work out where).

I was thinking of an approach where this PR is merged but with an additional check in the sip code that will raise a python exception `if ( !lastError().isNull() )`.

nyalldawson commented:
> I was thinking of an approach where this PR is merged but with an additional check in the sip code that will raise a python exception if ( !lastError().isNull() )

Sounds reasonable!

m-kuhn commented:
... and technically feasible without writing custom sip code for every single method QgsGeometry?

nyalldawson commented:
> ... and technically feasible without writing custom sip code for every single method QgsGeometry?

I don't know... is it? ;)

Personally... for the short term future I'm going to run in the opposite direction from any work involving sip mangling.
